### PR TITLE
MMCA-4972 : Implement search by a specific payment and View transactions on the Cash Account dashboard

### PR DESCRIPTION
**Pull request by HariHmrc**

### CODE REVIEW

File: app/controllers/CashAccountPaymentSearchController.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                                                   paymentSearchView: cash_account_payment_search,
+                                                   searchRepository: CashAccountSearchRepository
+                                                )(implicit executionContext: ExecutionContext,
+                                                  appConfig: AppConfig
+                                                ) extends FrontendController(mcc) with I18nSupport with Logging {
+
+  def search(searchValue: String,
+             page: Option[Int]): Action[AnyContent] = (authenticate andThen verifyEmail).async { implicit request =>
+
+    apiConnector.getCashAccount(request.eori).flatMap {
```
Comment (AdamCoops) [inline app/controllers/CashAccountPaymentSearchController.scala:50]: is it worth adding some spaces into this .flatmap cases just to increase readability

File: app/views/components/PaymentSearchResultHeaderV2.scala
Comment (tyagikumarvineet) [inline app/views/components/PaymentSearchResultHeaderV2.scala:1]: Better to rename the file name and class name to PaymentSearchResultHeader as we do not have old PaymentSearchResultHeader so no need to create V2

File: app/views/components/payment_search_results_v2.scala.html
Comment (tyagikumarvineet) [inline app/views/components/payment_search_results_v2.scala.html:1]: Please rename the file name to payment_search_results

File: conf/app.routes [hunk 2]
```diff
@@ -44,3 +44,5 @@ GET         /csv-download-error                     controllers.DownloadCsvContr
 
 GET         /transaction/:ref                       controllers.DeclarationDetailController.displayDetails(ref: String, page: Option[Int])
 GET         /transaction-search/:searchInput        controllers.DeclarationDetailController.displaySearchDetails(page: Option[Int], searchInput: String)
```
Comment (tyagikumarvineet) [inline conf/app.routes:46]: Why this has not been updated to POST method (to hide the search parameter i.e. MRN and UCR) as we discussed?

File: test/controllers/CashAccountPaymentSearchControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        val request =
+          FakeRequest(GET, routes.CashAccountPaymentSearchController.search(PAYMENT_SEARCH_VALUE, Some(1)).url)
+            .withSession("eori" -> eori)
+
+        val result = route(app, request).value
+        status(result) mustEqual OK
+      }
+    }
+
+    "return NOT Found view when search results are not available in cache repository" in new Setup {
```
Comment (tyagikumarvineet) [inline test/controllers/CashAccountPaymentSearchControllerSpec.scala:63]: Could you please add a test case for below values? Also update NOT _FOUND to the correct view as mentioned in earlier comments

when value is successfully retrieved but paymentTransfersList.paymentsWithdrawalsAndTransfers is None

File: test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala
Comment (tyagikumarvineet) [inline test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala:1]: Please update the class and file name as per comments earlier

File: test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import utils.SpecBase
+
+class CashTransactionsPaymentSearchEncrypterSpec extends SpecBase {
+
+  private val cipher = new AesGCMCrypto
+  private val encrypter = new CashAccountPaymentSearchEncrypter(cipher)
+  private val secretKey = "VqmXp7yigDFxbCUdDdNZVIvbW6RgPNJsliv6swQNCL8="
+
+
+  "encrypt / decrypt cashTransactions" must {
```
Comment (tyagikumarvineet) [inline test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala:34]: Better to update this test description to "encrypter" must {

File: test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    "encrypt and decrypt declaration search results correctly" in new Setup {
```
Comment (tyagikumarvineet) [inline test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala:36]: Better to rename it to as below as we are dealing with response details

"encrypt and decrypt declaration search response details correctly"

File: test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      val encryptedValue: EncryptedValue = encrypter.encryptSearchResponseDetail(
+        declarationsSearchResponseDetail, secretKey)
+
+      val decryptedObject: CashAccountTransactionSearchResponseDetail = encrypter.decryptSearchResponseDetail(
+        encryptedValue, secretKey)
+
+      decryptedObject mustEqual declarationsSearchResponseDetail
+    }
+
+    "encrypt and decrypt payment search results correctly" in new Setup {
```
Comment (tyagikumarvineet) [inline test/crypto/CashTransactionsPaymentSearchEncrypterSpec.scala:47]: Better to rename it to as below as we are dealing with response details

"encrypt and decrypt payment search response details correctly"

File: test/utils/TestData.scala [hunk 7]
```diff
@@ -85,4 +88,39 @@ object TestData {
   val PAGE_1000 = 1000
 
   val TEST_ID = "testId"
+
+  val LENGTH_40: Int = 40
+
+  val PAYMENT_SEARCH_VALUE: String = "$123.12"
```
Comment (tyagikumarvineet) [inline test/utils/TestData.scala:94]: Better to replace $ with £ as that is our main use case here

File: test/utils/UtilsSpec.scala [hunk 8]
```diff
@@ -89,6 +89,28 @@ class UtilsSpec extends SpecBase {
...
+
+  "buildCacheId" should {
+    "return a new cache id with given inputs" in {
+      buildCacheId("can1234", "$40.00") mustBe "can1234_$40.00"
+      buildCacheId("can1234", "-$20.10") mustBe "can1234_-$20.10"
+    }
+  }
+
+  "extractNumericValue" should {
+    "return only numeric value without pound or any symbols except minus" in {
```
Comment (tyagikumarvineet) [inline test/utils/UtilsSpec.scala:107]: Could you please add one assertion for 4 digits (before decimal point) and one for 5 digits?

Also better to add an assertion for the empty string

File: test/viewmodels/PaymentSearchResultStatementsViewModelSpec.scala [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  "apply" should {
+
+    "return viewModel object with correct content" when {
+
+      "cash transactions are available" in new Setup {
+        val viewModel01: PaymentSearchResultStatementsViewModel =
+          PaymentSearchResultStatementsViewModel(SEQ_OF_PAYMENT_DETAILS_01, None)
+
+        viewModel01.dailyStatements.size mustBe 3
```
Comment (tyagikumarvineet) [inline test/viewmodels/PaymentSearchResultStatementsViewModelSpec.scala:35]: Could you please add an assertion for the dailyStatements values along with this size assertion? Leave it if it takes time for you

File: test/viewmodels/PaymentSearchResultsV2ViewModelSpec.scala [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+class PaymentSearchResultsV2ViewModelSpec extends SpecBase {
+
+  "apply method" should {
+
+    "return correct contents without pagination" when {
+
+      "search results are less than records per page in config" in new Setup {
+
+        val viewModel = PaymentSearchResultsViewModel.apply(searchValue = PAYMENT_SEARCH_VALUE,
```
Comment (tyagikumarvineet) [inline test/viewmodels/PaymentSearchResultsV2ViewModelSpec.scala:38]: Add return type as its a public member. No need to explicitly mention .apply. Better to refactor as below. Same for line 55

        val viewModel: PaymentSearchResultsViewModel =
          PaymentSearchResultsViewModel(
            searchValue = PAYMENT_SEARCH_VALUE,
            account = cashAccount,
            paymentsWithdrawalsAndTransfers = SEQ_OF_PAYMENT_DETAILS_01,
            pageNo = Some(1))

File: test/connectors/CustomsFinancialsApiConnectorSpec.scala
Comment (tyagikumarvineet) [inline test/connectors/CustomsFinancialsApiConnectorSpec.scala:1]: This file is generating scalastyle warning as the max no of lines have been increased beyond 900. Increase the parameter named maxFileLength value to 950 in test-scalastyle-config.xml as the refactoring of this class will be done as part of separate ticket

File: conf/app.routes (hunk 2 above)
Comment (HariHmrc) [reply conf/app.routes:46]: I had to revoke it back to GET as merge conflicts were quite messy.

Comment (tyagikumarvineet) [reply conf/app.routes:46]: Okay, but have this been conveyed to Harsha to address it as part of separate ticket?

File: test/viewmodels/PaymentSearchResultsV2ViewModelSpec.scala
Comment (tyagikumarvineet) [inline test/viewmodels/PaymentSearchResultsV2ViewModelSpec.scala:1]: Could you please rename the file and the class to PaymentSearchResultsViewModelSpec as it should be aligned with its class?

File: conf/app.routes (hunk 2 above)
Comment (tyagikumarvineet) [reply conf/app.routes:46]: It will be handled as part of separate ticket and has been conveyed to Harsha.

**Review by tyagikumarvineet — APPROVED** [review]

**Review by AdamCoops — APPROVED** [review]

**Pull request merged**